Option to announce current song

bokherus opened the pull request:
Closes #342 

### Changes
- Send an embeded message to announce current playing song.
- Allow configuration of announcement per guild. (defaults to false)

### Embeded message style

<img width="417" alt="Screen Shot 2565-01-04 at 23 27 18" src="https://user-images.githubusercontent.com/9125776/148091162-c43f9990-ff23-4c64-b5b3-d9b20327b4e0.png">


### Configuration 

<img width="471" alt="Screen Shot 2565-01-04 at 23 26 24" src="https://user-images.githubusercontent.com/9125776/148091246-b3c747a3-23d9-48a9-8740-216e28f55099.png">


- [ ] I updated the changelog

codetheweb commented:
Thanks for putting this together.

I'm currently working on migrating Muse to slash commands, where this might be obsolete / wouldn't work (assuming Muse only has permissions to reply to commands--we can give Muse permission to create messages as well I guess if necessary).

Could you check out https://github.com/codetheweb/muse/issues/331#issuecomment-1001871013 and see what you think? The main thing missing from that would be notifying the user who requested the track.

bokherus commented:
I'll check it out. Looks great!

codetheweb commented:
I know the `.announceCurrentSong()` in Player is really short, but there's so little functional overlap between it and the rest of the Player's functionality that I think this would probably make more sense in its own service. You'd then want to make the Player emit events upon track change that you can listen to in your service (something like https://github.com/codetheweb/muse/pull/431/commits/0b20cb3982942196303dea3cbbb98644dde441a3#diff-56fde53d4e5107906947fc7476ccd5da900c3a183a34c8137a17080f55416669R34).

bokherus commented:
I have updated to use event emitter and tried to created as a service but then I changed my mind as it seems like an overkill because the functionality only need an event listener that has song context and channel to send the message. Please let me know what you think about this

codetheweb commented:
It's looking better to me so far. I think you probably want to listen to a track change event rather than play / pause.

bokherus commented:
I think it depends on the functionality we wanted. In this case, I listen to play events so that whenever user resumes a play from pause state it would also show the message since it doesn't switch track in this case.

bokherus commented:
@codetheweb I've fixed the conflict and update code to use prisma. There's some issue with the migration generation script, thanks to help from @pondpiu it's now resolved. I've also added the fix in this commit https://github.com/codetheweb/muse/pull/453/commits/efd10a9fbb86eede57dde932bd88185d8fbdee02

codetheweb commented on schema.prisma:
```diff
@@ -29,6 +29,7 @@ model Setting {
   channel       String?
   finishedSetup Boolean  @default(false)
   playlistLimit Int      @default(50)
+  announceSongs Boolean  @default(true)
```
```suggestion
  announceSongs Boolean  @default(false)
```

I think this should probably default to false so we don't change existing behavior.

codetheweb reviewed (changes requested):
Looking good.

I think I might have misunderstood the intent of this PR. Is the desired behavior that Muse only announces songs when starting playback? I was originally thinking that an announcement would be sent upon every track change but that seems fine.

> There's some issue with the migration generation script, thanks to help from @pondpiu it's now resolved.

Thank you. :)

bokherus commented:
Yeah I think it's the intent of the issue owner to announce the current playing song. I just thought it's logical for cases where user disconnect the bot without clearing the queue and then use play command to continue with that's left in the queue. I think this should be announced. Let's discuss about tweaks to match the concept you want to do with this bot. For track change it's a bit tricky because when the queue is empty and a song just got added and immediately start playback is this considered as a track changed or is it only a play event?

codetheweb commented:
Yes, letting people know what the current track is after `.p` without having to do `.queue` makes a lot of sense and something that should be core functionality and enabled by default. Announcing tracks on track change seems more like an addon and something that should probably be disabled by default.

That being said, it would probably make more sense to focus on the first use case for this PR and open a second one if you want to add support for the second use case as well.

I think it'd probably work better for the first use case if the functionality was moved into the play command handler rather than listening to player events (`src/commands/play.ts`). This will also make it easier to migrate to slash commands. Probably should also remove `Settings.annouceSongs` and just always do it (since, at least in my mind, that setting is referring to the second use case).

Feel free to open a new PR if that seems easier.

Sorry for all the back-and-forth, I should have clarified how this works in the beginning.

bokherus commented:
> Yes, letting people know what the current track is after `.p` without having to do `.queue` makes a lot of sense and something that should be core functionality and enabled by default.

I'm not sure if I understand this correctly. Do you mean that whenever `play` command is used it will show what's the current playing tracks is? Currently, I'm considering 3 scenario when `play` command is used. 
1. No song in queue (song is then added and immediately played)
2. Existing song in queue but not playing (resume playback)
3. Existing song in queue and is playing (add new song to queue)

**Here are my opinions for handling immediate response after the command is received and executed**
Scenario 1 - show what is being played (immediate feedback to see if we get the correct song in case of text search)
Scenario 2 - show what song is going to be play since we might lose context (pause/resume on different day etc.)
Scenario 3 - responding with what song is added to the queue.

Announcing song on each track change or playback is however not bound to `play` command as it's more like a notification than a response to command.

I take a look into some music bots to get some example. Please take a look at the screenshots below. In my opinion, I like the flow on pancake bot for queuing song and playback notification, `skip` command is also better than probot as it's showing what's playing and not what's skipped. However, pause and resume on probot seems to be better for me as it's showing the song. Please let me know what you think should be the behavior of the bot in these scenarios.

![image](https://user-images.githubusercontent.com/9125776/148652601-aa17ee39-6a54-4fb8-8821-6e78e1b0b302.png)
<img width="424" alt="Screen Shot 2565-01-08 at 23 48 44" src="https://user-images.githubusercontent.com/9125776/148652607-ed4326f9-dc32-46dd-8a65-34ac2ecc7d28.png">
<img width="676" alt="Screen Shot 2565-01-08 at 23 50 01" src="https://user-images.githubusercontent.com/9125776/148652622-7fd898f2-33fe-47ba-a7a8-79b8512c3eda.png">

codetheweb commented:
Thanks for further clarifying.

I like the design of pancake bot's responses in the sense that you know if the track you added is going to be currently playing or added to the back of the queue. I don't really like that the embeds look so different for the same action. The response after skipping a track on pancake bot also makes sense.

I think (if you want to go the embed route) that a universal embed design for play, pause, and skip commands would be great. It should also let the user know what position in the queue their track was added, but again I wouldn't want the design in that scenario to be significantly different (as is the case with pancake bot).

bokherus commented:
I've created a new PR https://github.com/codetheweb/muse/pull/470. Closing this one.
